plutus-contract-exe/src/Language/Plutus/Contract/Transaction.hs

```diff
@@ -0,0 +1,171 @@
+
+-- | Create an `UnbalancedTx` that pays money to a script address.
+payToScript :: Value -> Address -> DataScript -> UnbalancedTx
+payToScript v a ds = unbalancedTx [] [outp] where
+    outp = Tx.scriptTxOut' v a ds
+
+-- | Create an `UnbalancedTx` that collects script outputs from the
+--   address of the given validator script, using the same redeemer script
+--   for all outputs. See 'Wallet.API.collectFromScript'
+collectFromScript
+    :: AddressMap
+    -> ValidatorScript
+    -> RedeemerScript
+    -> UnbalancedTx
+collectFromScript = collectFromScriptFilter (\_ -> const True)
+
+-- | See 'Wallet.API.collectFromScriptFilter'
+collectFromScriptFilter
+    :: (TxOutRef -> TxOut -> Bool)
+    -> AddressMap
+    -> ValidatorScript
+    -> RedeemerScript
+    -> UnbalancedTx
+collectFromScriptFilter flt am vls red =
+    let utxo    = fromMaybe Map.empty $ am ^. at (Tx.scriptAddress vls)
+        ourUtxo = Map.toList $ Map.filterWithKey flt utxo
+        mkTxIn (ref, txo)   = (Tx.scriptTxIn ref vls red, view Tx.outValue txo)
+        txInputs  = mkTxIn <$> ourUtxo
+    in
+    unbalancedTx txInputs []
+
+{- note [Unbalanced transactions]
+
+To turn an 'UnbalancedTx' into a valid transaction that can be submitted to the
+network, the contract backend needs to
+
+* Balance it.
+  If the total value of `utxInputs` + the `txForge` field is
+  greater than the total value of `utxOutput`, then one or more public key
+  outputs need to be added. How many and what addresses they are is up
+  to the wallet (probably configurable).
+  If the total balance `utxInputs` + the `txForge` field is less than
+  the total value of `utxOutput`, then one or more public key inputs need
+  to be added (and potentially some outputs for the change)
+
+* Compute fees.
+  Once the final size of the transaction is known, the fees for the transaction
+  can be computed. The transaction fee needs to be paid for with additional
+  inputs so I assume that this step and the previous step will be combined.
+
+  Also note that even if the 'UnbalancedTx' that we get from the contract
+  endpoint happens to be balanced already, we still need to add fees to it. So
+  we can't skip the balancing & fee computation step.
+
+* Sign it.
+  The signing process needs to provide signatures for all public key
+  inputs in the balanced transaction, and for all public keys in the
+  `utxRequiredSignatures` field.
+
+While there is an 'empty' transaction we can't make 'UnbalancedTx' a monoid
+because it's not clear what the binary operator should do with the validity
+interval. There are two valid options: Hull and intersection. ('always' is the
+unit for the intersection but then there is the issue that we don't have a
+canonical representation of the empty interval (that's why 'intersection'
+returns a 'Maybe Interval'.))
+
+TODO:  I think a good way forward would be to make
+'Interval' a lattice, and then make 'UnbalancedTx' a lattice too.
```

Comment:
I strongly agree with making `Interval` a lattice. It's easy to add an empty interval (even using picking any interval with `l >h` would do).

I'm not sure that `UnbalancedTx` should be a lattice. The operator where you use the monoid on the other fields and `intersection` on the interval seems useful, and seems most naturally to be `join` (since it "does both"), I'm not sure what `meet` would be.

Reply:
> even using picking any interval with `l >h` would do

But how should we define `empty :: Interval a` ?

Reply:
Right, you probably need something like 
```
class Lattice a => BoundedLattice a where
    top :: a
    bottom :: a

instance BoundedLattice a => BoundedLattice (Interval a) where
    top = Interval bottom top
    bottom = Interval top bottom
````

Possibly we should make it possible to specify the closure of the endpoints in an interval since I think you probably want top to be `[bottom, top]` and bottom to be `(top, bottom)` so they work well in the degenerate case with one element.

Reply:
https://github.com/input-output-hk/plutus/issues/1309

Reply:
https://github.com/input-output-hk/plutus/issues/1309